bin/convert-env.sh

```diff
@@ -17,39 +17,36 @@ then
 fi
 
 # shape old env vars into app-server compatible ones
-# mediation layer
-if [ -z "$ZWED_node_mediationLayer_server_gatewayPort" ]
-then
-  if [ -n "$GATEWAY_PORT" ]
-  then
-    export ZWED_node_mediationLayer_server_gatewayPort=$GATEWAY_PORT
+# mediation layer - gateway
+if [ -z "$ZWED_node_mediationLayer_server_gatewayHostname" ]; then
+  if [ -n "$ZWE_EXTERNAL_HOSTS" ]; then
+    #just the first value in the csv
+    export ZWED_node_mediationLayer_server_gatewayHostname=$(echo "${ZWE_EXTERNAL_HOSTS}" | head -n1 | sed "s/,/ /g" | cut -d " " -f1)
+  elif [ -n "$ZOWE_EXPLORER_HOST" ]; then
+    export ZWED_node_mediationLayer_server_gatewayHostname=$ZOWE_EXPLORER_HOST
   fi
 fi
-
-if [ -z "$ZWED_node_mediationLayer_server_port" ]
-then
-  if [ -n "$DISCOVERY_PORT" ]
-  then
-    export ZWED_node_mediationLayer_server_port=$DISCOVERY_PORT
-  fi
+if [ -z "$ZWED_node_mediationLayer_server_gatewayPort" -a -n "$GATEWAY_PORT" ]; then
+  export ZWED_node_mediationLayer_server_gatewayPort=$GATEWAY_PORT
 fi
-
-if [ -z "$ZWED_node_mediationLayer_server_hostname" ]
-then
-  if [ -n "$ZOWE_EXPLORER_HOST" ]
-  then
-    export ZWED_node_mediationLayer_server_hostname=$ZOWE_EXPLORER_HOST
-    if [ -n "$ZWED_node_mediationLayer_server_port" ]
-    then
-      case "$LAUNCH_COMPONENT_GROUPS" in
-        *GATEWAY*)
-          #All conditions met for app-server behind gateway: hostname, port, and component
-          export ZWED_node_mediationLayer_enabled="true"
-          ;;
-      esac
-    fi
+# mediation layer - discovery
+# try to extract discovery server from ZWE_DISCOVERY_SERVICES_LIST
+if [ -n "${ZWE_DISCOVERY_SERVICES_LIST}" -a -z "$ZWED_node_mediationLayer_server_hostname" -a -z "$ZWED_node_mediationLayer_server_port" ]; then
+  # get first discovery, there is chance the first discovery may fail
+  first_discovery=$(echo "${ZWE_DISCOVERY_SERVICES_LIST}" | awk -F"," '{print $1;}' | awk -F"/" '{print $3;}')
+  if [ -n "${first_discovery}" ]; then
+    export ZWED_node_mediationLayer_server_hostname=$(echo "${first_discovery}" | awk -F":" '{print $1;}')
+    export ZWED_node_mediationLayer_server_port=$(echo "${first_discovery}" | awk -F":" '{print $2;}')
+    export ZWED_node_mediationLayer_enabled="true"
```

Comment:
this appears to be the cause of a bug we're seeing where the app-server believes apiml exists even when it doesnt.
it appears to happen because ZWE_DISCOVERY_SERVICES_LIST has a value even if LAUNCH_COMPONENT_GROUPS doesnt contain gateway.

I understand that in HA, the discovery service will not always exist in the same HA node as the app-server, but the app-server still needs some value to know whether or not apiml exists, and this doesn't seem to get set to undefined in the case that apiml doesnt exist. At least, that's what we see in build zowe-install-packaging build 2049.

Reply:
Yes I'm a little confused about this configuration. Does that mean enablement of Gateway or Discovery, or both?

In HA deployment, I originally adding a comment to `LAUNCH_COMPONENT_GROUPS` saying this is deprecated with YAML config. Because the user has more flexibility to have different combination, it's very hard to create a correct `LAUNCH_COMPONENT_GROUPS`. What HA case will really use is `ZWE_LAUNCH_COMPONENTS`, and this should always be generated based on YAML config.

I can try to calculate a value of `ZWED_node_mediationLayer_enabled` based on enablement of APIML across the whole cluster. But the problem is still:
- does this value mean for discovery so zlux can register itself?
- even though we see discovery/gateway are enabled across cluster, but there is chance the sysadm doesn't start that Zowe instance yet, or just plans to start it after the instance where app-server locates.

Please see if there are any good solution, I can fix this PR as well as the YAML/instance.env converter.

Reply:
zlux needs a value to know whether or not to try and contact the discovery server, and to start doing a heartbeat.
zlux also needs a value to know whether or not to try to use the gateway for login, instead of zss.

so, zlux has conditionals about whether or not apiml exists. I understand that HA can't use LAUNCH_COMPONENT_GROUPS and thats ok, but a different variable that serves the same purpose for zlux detection is needed. It's fine if the ZWE_DISCOVERY_SERVICE_LIST is used for that purpose but it must be set to undefined if the gateway or discovery servers arent going to be running. It's also fine if ZWE_LAUNCH_COMPONENTS serves that purpose, but I thought it did not because it only tells you about what exists in the current node.

It's fine to say that zlux needs to figure all this out at runtime rather than startup, but if that were the case then we need a more involved PR that will take some time to research, and could be a breaking change for extenders?